src/Microsoft.Unity.Analyzers/IndirectionMessage.cs

```diff
@@ -0,0 +1,107 @@
+/*--------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See LICENSE in the project root for license information.
+ *-------------------------------------------------------------------------------------------*/
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.Unity.Analyzers.Resources;
+
+namespace Microsoft.Unity.Analyzers
+{
+	[DiagnosticAnalyzer(LanguageNames.CSharp)]
+	public class IndirectionMessageAnalyzer : DiagnosticAnalyzer
+	{
+		internal static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(
+			id: "UNT0019",
+			title: Strings.IndirectionMessageDiagnosticTitle,
+			messageFormat: Strings.IndirectionMessageDiagnosticMessageFormat,
+			category: DiagnosticCategory.Performance,
+			defaultSeverity: DiagnosticSeverity.Info,
+			isEnabledByDefault: true,
+			description: Strings.IndirectionMessageDiagnosticDescription);
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+		public override void Initialize(AnalysisContext context)
+		{
+			context.EnableConcurrentExecution();
+			context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+			context.RegisterSyntaxNodeAction(AnalyzeExpression, SyntaxKind.SimpleMemberAccessExpression);
+		}
+
+
+		private static void AnalyzeExpression(SyntaxNodeAnalysisContext context)
+		{
+			if (!(context.Node is MemberAccessExpressionSyntax))
+				return;
+
+			var access = (MemberAccessExpressionSyntax)context.Node;
+			var symbol = context.SemanticModel.GetSymbolInfo(access);
+			var typeInfo = context.SemanticModel.GetTypeInfo(access.Expression);
+			
+			if (symbol.Symbol == null)
+				return;
+
+			if (!(symbol.Symbol is IPropertySymbol))
+				return;
+				
+			if (typeInfo.Type == null)
+				return;
+				
+			if (!typeInfo.Type.Extends(typeof(UnityEngine.GameObject)))
+				return;
+
+			if (access.Name.ToFullString() != "gameObject")
```

Comment:
Please move this between the declaration of `var access` and accessing the semantic model.

The reason for that is that accessing the SemanticModel is time confusing, whereas comparing the name of the member being access is really fast. This way we're only going to do the costly analysis for members named `gameObject`.